synthesizer/process/src/finalize.rs

```diff
@@ -361,39 +382,34 @@ fn finalize_transition<N: Network, P: FinalizeStorage<N>>(
 }
 
 // A helper struct to track the execution of a finalize block.
-struct FinalizeState<'a, N: Network> {
+struct FinalizeState<N: Network> {
     // A counter for the index of the commands.
     counter: usize,
-    // The finalize logic.
-    finalize: &'a Finalize<N>,
     // The registers.
     registers: FinalizeRegisters<N>,
     // The stack.
-    stack: &'a Stack<N>,
+    stack: Arc<Stack<N>>,
     // Call counter.
     call_counter: usize,
     // Awaited futures.
     awaited: HashSet<Register<N>>,
 }
 
 // A helper function to initialize the finalize state.
-fn initialize_finalize_state<'a, N: Network>(
+fn initialize_finalize_state<N: Network>(
     state: FinalizeGlobalState,
     future: &Future<N>,
-    stack: &'a Stack<N>,
+    stack: &Arc<Stack<N>>,
     transition_id: N::TransitionID,
     nonce: u64,
-) -> Result<FinalizeState<'a, N>> {
-    // Get the finalize logic and the stack.
-    let (finalize, stack) = match stack.program_id() == future.program_id() {
-        true => (stack.get_function_ref(future.function_name())?.finalize_logic(), stack),
-        false => {
-            let stack = stack.get_external_stack(future.program_id())?.as_ref();
-            (stack.get_function_ref(future.function_name())?.finalize_logic(), stack)
-        }
+) -> Result<FinalizeState<N>> {
+    // Get the stack.
+    let stack = match stack.get_external_stack(future.program_id()) {
```

Comment:
Since we're not checking if `stack.program_id() == future.program_id()` anymore, should we keep it as a lazy defense-in-depth `ensure!`?